packages/contracts/src/contracts/current/forwarder/MixinForwarderCore.sol

```diff
@@ -0,0 +1,147 @@
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        address token = readAddress(orders[0].makerAssetData, 0);
+        require(token == address(ZRX_TOKEN), TAKER_ASSET_ZRX_REQUIRED);
+        for (uint256 i = 0; i < orders.length; i++) {
+            // Token being bought by taker must be the same for each order
+            require(areBytesEqual(orders[i].makerAssetData, orders[0].makerAssetData), SAME_ASSET_TYPE_REQUIRED);
+
+            // Calculate the remaining amount of makerToken to buy
+            uint256 remainingMakerTokenFillAmount = safeSub(feeAmount, totalFillResults.makerAssetFilledAmount);
+
+            // Convert the remaining amount of makerToken to buy into remaining amount
+            // of takerToken to sell, assuming entire amount can be sold in the current order
+            uint256 remainingTakerTokenFillAmount = getPartialAmount(
+                orders[i].takerAssetAmount,
+                safeSub(orders[i].makerAssetAmount, orders[i].takerFee), // our exchange rate after fees 
+                remainingMakerTokenFillAmount);
+
+            // Attempt to sell the remaining amount of takerToken
+            // Round up the amount to ensure we don't under buy by a fractional amount
+            Exchange.FillResults memory singleFillResult = EXCHANGE.fillOrder(
+                orders[i],
+                safeAdd(remainingTakerTokenFillAmount, 1),
+                signatures[i]
+            );
+
+            // We didn't buy the full amount when buying ZRX as some were taken for fees
+            singleFillResult.makerAssetFilledAmount = safeSub(singleFillResult.makerAssetFilledAmount, singleFillResult.takerFeePaid);
+            // Update amounts filled and fees paid by maker and taker
+            addFillResults(totalFillResults, singleFillResult);
+
+            // Stop execution if the entire amount of makerToken has been bought
+            if (totalFillResults.makerAssetFilledAmount >= feeAmount) {
+                break;
+            }
+        }
+        return totalFillResults;
+    }
+
+    /// @dev Checks whether the amount of tokens sold against the amount of tokens requested
+    ///      is within a certain threshold. This ensures the caller gets a fair deal when
+    ///      performing any token fee abstraction. Threshold is 95%. If fee abstraction costs more than
+    ///      5% of the total transaction, we return false.
+    /// @param requestedTokenAmount The amount the user requested, or sent in to a payable function
+    /// @param soldTokenAmount The amount of the token that was sold after fee abstraction
+    /// @return bool of whether this is within an acceptable threshold
+    function isAcceptableThreshold(uint256 requestedTokenAmount, uint256 soldTokenAmount)
+        internal
+        pure
+        returns (bool)
+    {
+        uint256 acceptableProportion = safeDiv(safeMul(requestedTokenAmount, ALLOWABLE_EXCHANGE_PERCENTAGE), PERCENTAGE_DENOMINATOR);
```

Comment:
I've seen this in a few places, probably worth mentioning that we do have a `getPartialAmount` now in `LibMath`.